File: Lv2/다리를 지나는 트럭.cpp
#include <string>
#include <vector>
#include <queue>
#include <iostream>

using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {
    int answer = 0, truck_count = 0, total_weight = 0;
    queue<int> q;
    
    while (truck_count < truck_weights.size()) {
        if (q.size() == bridge_length) {
            int poped = q.front();
            q.pop();
            if (poped != 0)
                total_weight -= poped;
        }
        
        
        if (total_weight + truck_weights.at(truck_count) <= weight && q.size() < bridge_length) {
            total_weight += truck_weights.at(truck_count);
            q.push(truck_weights.at(truck_count++));
        }
        else
            q.push(0);
        
        answer++;
    }
     
    answer += bridge_length;
    
    return answer;
}

File: Lv2/다리를 지나는 트럭.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    answer = 0
    total_weight = 0
    truck_count = 0
    bridge = deque()
    
    while truck_count != len(truck_weights): 
        if len(bridge) == bridge_length:
            total_weight -= bridge.popleft()
        
        if total_weight + truck_weights[truck_count] <= weight:
            bridge.append(truck_weights[truck_count])
            total_weight += truck_weights[truck_count]
            truck_count += 1
        else:
            bridge.append(0)
        
        answer += 1
    
    return answer + bridge_length
